### PR TITLE
[BugFix] Fix bug of regularizer in tf.layers under DistributionStrategy

### PR DESCRIPTION
**Pull request by wangsiyu**
There is a bug when building graph in tf.layers under DistributionStrategy. The regularizer subgraph will be only built on `replica 0` and it will be ignored on other `replicas` in `MirroredStrategy`.  Actually, regularizer subgraph should be related to variable type. 
Another, if one builds regularization_loss via tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSS), it may cause some problems:
1. Regroup Assertion error: 
    If one builds two layers with regularizers but only trained with first layer，and get regularization loss from global collection under `MirroredStrategy`, the second layer will has gradient on replica 0 but None on other replicas, which leads to regroup error after call into `merge_call`(Although the second layer's regularization is unnecessary to build, but it is not easy for users to remove in model_fn).
2. Performance issue: 
    Global collection is not thread local in `model_fn`, so we will get all regularizers on all replicas. However, before we fix this bug, the regularizer will only exist on `replica 0`. This will involve device communications. 

Please take a look at this PR. Thanks. @yuefengz @martinwicke @anj-s

### CODE REVIEW

**Comment by guptapriya** [comment]
Thanks for the PR! Looks like the unit test file is missing, can you add that as well before we review? thanks

**Comment by wangsiyu** [comment]
@guptapriya Sorry, I missed a source file. Now the unit test has been added. Please check it again. Thanks.

**Comment by guptapriya** [comment]
Thanks for adding the test file! I'll let @anj-s review.

**Comment by wangsiyu** [comment]
This PR has been pending for some days. Any update for me？

File: tensorflow/python/layers/base.py [hunk 1]
```diff
@@ -450,9 +453,61 @@ def _should_add_regularizer(variable, existing_variable_set):
...
 
         if regularizer:
-          if (ops.executing_eagerly_outside_functions()
-              or _should_add_regularizer(variable, existing_variables)):
-            self._handle_weight_regularization(name, variable, regularizer)
+
+          # In order to support DistributionStrategy, we move regularization
+          # in the following `add_regularizer` merge_fn. Regularization should
+          # be declared according to variable type. For example, we should call
+          # regularizer on all repilcas when it is a `MirroredVariable`.
```
Comment (anj-s) [inline tensorflow/python/layers/base.py]: s/repilcas/replicas

File: tensorflow/python/layers/base.py [hunk 2]
```diff
@@ -450,9 +453,61 @@ def _should_add_regularizer(variable, existing_variable_set):
...
+          # However, `MirroredVariable` is created when first replica thread is
+          # called, which makes regularizer statements on other replica
+          # thread ignored because of the `existing_variabale` check. Therefore,
+          # it is reasonable to put regularization statement in
+          # `cross_replica` context.
```
Comment (anj-s) [inline tensorflow/python/layers/base.py]: I am not convinced that this is the solution. Won't the value of the regularization loss be the same on all replicas because it is a function of the variable which is mirrored?

File: tensorflow/python/distribute/regularization_loss_test.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  v1_value, v2_value = sess.run([v1, v2])
+  saver = saver_lib.Saver()
+  saver.save(
+      sess,
+      checkpoint_prefix,
+      global_step=0,
+      latest_filename=checkpoint_state_name)
+  return v1_value, v2_value
+
+class RegularizationWithDistributionStrategyTest(
```
Comment (anj-s) [inline tensorflow/python/distribute/regularization_loss_test.py]: Is this test representative of the bug that you ran into?

File: tensorflow/python/layers/base.py [hunk 4]
```diff
@@ -450,9 +453,61 @@ def _should_add_regularizer(variable, existing_variable_set):
...
+
+              g = ops.get_default_graph()
+              l_before = g.get_collection(ops.GraphKeys.REGULARIZATION_LOSSES)
+              for d in variable.devices:
+                self._handle_weight_regularization(
+                    name, variable.get(d), regularizer.get(d))
+
+              # Since the regularization loss operations have been added to
+              # global collection, so we manually remove them and replace with
+              # the `Mirrored` type.
```
Comment (anj-s) [inline tensorflow/python/layers/base.py]: I am not sure I understand why this needs to be done. can you give me an example of your use case for adding this?

**Review by anj-s — CHANGES_REQUESTED** [review]

File: tensorflow/python/layers/base.py (hunk 2 above)
Comment (wangsiyu) [reply tensorflow/python/layers/base.py]: Thanks for your comments. The problem is In current `MirroredStrategy`, regularization loss will not be declared with `MirroredVariable` at the same time. You should pay attention to the `existing_variable`, it contains `MirroredVariable` fetched from tf.global_variables. So, the first replica will create `MirroredVariable` and put them into tf.global_variables list. And the second replica will see all `MirroredVariable` in `existing_variables` because they have been put by first replica! Therefore, first replica does have regularization_loss, but other replicas does not.

It has performance issue in the following code in `Estimator`, the performance will be bad.
```
def model_fn():
    regularization_losses = tf.get_collection(
              tf.GraphKeys.REGULARIZATION_LOSSES)
    regularization_loss = tf.add_n(regularization_losses)
```
Of course the `model_fn` will be called on each replica to construct reg_loss and the result is right, but the regularization nodes are placed on replica 0, so other replicas will fetch them from replica 0 which involve unnecessary memcpy. 

The reason I put them in the `merge_call` is that I want to declare multiple `regularizer` on multiple devices. If I don't put them into `merge_call`, other replicas will loss the chance to declare their own `regularizer` because of the `MirroredVariable` in `existing_variables`.

I know the `global_collections` will be deprecated in TF2.0, but still some users are eager to use 1.0 api like tf.layers. That's the reason why I fix this bug in this way.

File: tensorflow/python/layers/base.py (hunk 4 above)
Comment (wangsiyu) [reply tensorflow/python/layers/base.py]: Like `MirroredVariable`, the `regularizer` should be formated in `DistributedValues` not the atomic type.  So If I get values from  `tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES)`, I can get its value on current device.

File: tensorflow/python/distribute/regularization_loss_test.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+from tensorflow.python.distribute import combinations
+from tensorflow.python.distribute import strategy_combinations
+from tensorflow.python.distribute import values as value_lib
+from tensorflow.python.framework import ops
+from tensorflow.python.layers import base as base_layers
+from tensorflow.python.ops import math_ops
+from tensorflow.python.platform import test
+
+
+def _create_checkpoints(sess, checkpoint_dir):
```
[the remark this answers is not in the document]
Comment (wangsiyu) [reply tensorflow/python/distribute/regularization_loss_test.py]: This should be deleted. Thanks for your comments.

File: tensorflow/python/distribute/regularization_loss_test.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+              strategy_combinations.mirrored_strategy_with_two_gpus,
+          ],
+          mode=["graph"]))
+  def testMoreThanOneReplicasWithLayers(self, distribution):
+    with distribution.scope():
+      distribution.extended.call_for_each_replica(self.model_fn)
+      reg_list = ops.get_collection(ops.GraphKeys.REGULARIZATION_LOSSES)
+      self.assertEqual(len(reg_list), 1)
+      for reg_loss in reg_list:
+        self.assertTrue(isinstance(reg_loss, value_lib.Mirrored))
```
[the remark this answers is not in the document]
Comment (wangsiyu) [reply tensorflow/python/distribute/regularization_loss_test.py]: Good point.

File: tensorflow/python/layers/base.py (hunk 2 above)
Comment (wangsiyu) [reply tensorflow/python/layers/base.py]: Thanks for your comments @anj-s . The problem is In current MirroredStrategy, regularizer will not be declared with `MirroredVariable` declaration at the same time. You should pay attention to the existing_variable, it contains MirroredVariable fetched from tf.global_variables. So, the first replica will create MirroredVariable and put them into tf.global_variables list. And the second replica will see all MirroredVariable in existing_variables because they have been put by first replica! Therefore, first replica does have regularization_loss, but other replicas does not.

It has performance issue in the following code in Estimator, the performance will be bad.

def model_fn():
    regularization_losses = tf.get_collection(
              tf.GraphKeys.REGULARIZATION_LOSSES)
    regularization_loss = tf.add_n(regularization_losses)
Of course the model_fn will be called on each replica to construct reg_loss and the result is right, but the regularization nodes are placed on replica 0, so other replicas will fetch them from replica 0 which involve unnecessary memcpy.

The reason I put them in the merge_call is that I want to declare multiple regularizer on multiple devices. If I don't put them into merge_call, other replicas will loss the chance to declare their own regularizer because of the MirroredVariable in existing_variables.

I know the global_collections will be deprecated in TF2.0, but still some users are eager to use 1.0 api like tf.layers. That's the reason why I fix this bug in this way.

File: tensorflow/python/layers/base.py (hunk 4 above)
Comment (wangsiyu) [reply tensorflow/python/layers/base.py]: Like MirroredVariable, the regularizer should be formated in DistributedValues not the atomic type. So If I get values from tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES), I can get its value on current device.

**Comment by wangsiyu** [comment]
The coding style I will refine later. Thanks.

File: tensorflow/python/distribute/regularization_loss_test.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      combinations.combine(
+          distribution=[
+              strategy_combinations.default_strategy,
+              strategy_combinations.one_device_strategy,
+          ],
+          mode=["graph"]))
+  def testOneReplicaWithLayers(self, distribution):
+    with distribution.scope():
+      distribution.extended.call_for_each_replica(self.model_fn)
+      reg_list = ops.get_collection(ops.GraphKeys.REGULARIZATION_LOSSES)
```
[the remark this answers is not in the document]
Comment (wangsiyu) [reply tensorflow/python/distribute/regularization_loss_test.py]: Is this necessary to make this change？ I found the checkpointing_test.py also use `call_for_each_replica`.

**Comment by wangsiyu** [comment]
@anj-s code has been refined. Please check it again. Thanks.

File: tensorflow/python/distribute/regularization_loss_test.py (hunk 3 above)
Comment (wangsiyu) [reply tensorflow/python/distribute/regularization_loss_test.py]: Yes, The point is to count the regularizer op numbers.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by karmel** [comment]
Thanks, @wangsiyu -- we have discussed this contribution a fair amount internally, and unfortunately, we cannot accept these changes to tf.layers. The tf.layers APIs have been deprecated, and we can only accept critical bug fixes for supported usages. 

This change has far-reaching effects for tf.layers (eg the addition of mirrored variables to a collection), and we cannot fully anticipate the effects of those changes on existing users. Further, Distribution Strategies with Estimator is experimental currently and not fully supported, and Distribution Strategies with Estimator with tf.layers is even further from the preferred usage, so rather than make changes that could affect current users, we would encourage you to take one of two approaches:

1. If you want to use Distribution Strategies, please consider migrating to 2.0 APIs-- in this case, tf.keras.layers rather than tf.layers. The Keras Layers have close analogues to each tf.layers layer, and so you can convert one-by-one to the new Layers without changing model architecture significantly.

2. Update the model_fn to explicitly add L2 loss. This is what we do in the Resnet example here: https://github.com/tensorflow/models/blob/master/official/r1/resnet/resnet_run_loop.py#L428

Thanks for your understanding here.

**Comment by rthadur** [comment]
thank you for your contribution , closing this PR based on Karmel comments.
